### PR TITLE
feat(evals): add per-metric objective support to runner (closes #1091)

### LINKED ISSUE #1091: [FEATURE] semantica.evals: per-metric objective support (AIP Evals-style direction + threshold)
## Problem Statement

`semantica.evals` (implemented in #1090) runs named evaluators and aggregates per-case pass/fail, but the **pass judgement is hard-coded inside each evaluator** — a higher score always means better. There is no way to express an evaluation objective, e.g. "lower edit distance is better" or "F1 must be at least 0.7", at the run level.

For domain-specific benchmarks (the `semantica.evals` harnesses mentioned in `docs/community-projects.md`), teams need to apply different thresholds or directions to the same metric without forking the evaluator. Palantir AIP Evals models this explicitly: each metric carries an **objective** (for Boolean metrics, the expected value `true`/`false`; for numeric metrics, a `maximize`/`minimize` direction and an optional threshold), and a test case passes when **all** its metrics meet their objectives.

## Proposed Solution

Add a per-metric **objective** layer to the `evaluate()` runner, aligned with AIP Evals:

- Config: `config={"<evaluator_name>": {"objective": {"direction": "maximize"|"minimize", "threshold": 0.7}}}` or `{"objective": {"direction": "minimize"}}` (threshold optional).
- The runner re-evaluates each metric's pass/fail against the objective, overriding the evaluator's built-in default when an objective is present.
- Backward compatible: no objective configured → behavior unchanged (evaluator's own `passed` verdict is used).
- The aggregate `pass_rate` / case `status` semantics are unchanged (a case passes iff all metrics pass; error still wins over fail).

```python
from semantica.evals import evaluate

# "closer to 0 is better" — default levenshtein (>=0.8) is overridden
result = evaluate(
    [("night", "nacht")],
    evaluators=["levenshtein"],
    config={"levenshtein": {"objective": {"direction": "minimize", "threshold": 0.5}}},
)
```

## Alternatives Considered

1. **Change evaluators to return objective-aware metrics** — invasive, breaks the simple `fn(actual, expected, config) -> EvalMetric` contract for every evaluator.
2. **New `evaluate_with_objectives()` API** — duplicates the runner; keeps the layers apart but adds a second entry point to document and test.
3. **Objective only as threshold (no direction)** — covers half the use case; AIP Evals supports direction, and benchmarks genuinely need both.

Recommended: option 1 in the Proposed Solution (runner-layer override), which is the smallest surface change and matches how AIP Evals configures objectives per evaluation function.

## Use Cases

1. Domain benchmark harnesses (precision/recall) that apply stricter thresholds for production vs. experimental runs without touching evaluator code.
2. Cost-sensitive similarity checks where a **lower** distance is better and the default direction is wrong for the metric.
3. Regression gates: assert "F1 ≥ 0.7" or "error rate ≤ 0.05" as run-level criteria.

## Impact Assessment

- **Who would benefit?** Teams building evaluation harnesses on `semantica.evals`; anyone comparing metric behavior across runs.
- **Priority**: Medium
- **Breaking Changes**: No — additive, backward-compatible config.
- **Dependencies**: None; builds on the runner introduced in #1090.

## Implementation Ideas

- Runner reads `config["<name>"]["objective"]`; applies `direction`/`threshold` to `metric.score` and overwrites `metric.passed` (or surfaces an adjusted metric) before aggregation.
- New tests in `tests/evals/test_runner.py` for: minimize+threshold, maximize+threshold, no-threshold direction, no-objective fallback (unchanged behavior), interaction with error metrics.

## Additional Context

- Related PR: #1090 (the evals module this builds on).
- Reference: Palantir AIP Evals — "When configuring an evaluator, you can define an objective for each metric. For Boolean metrics, specify whether the metric should be `true` or `false`. For numeric metrics, set the direction of optimization; specify either a maximize (higher values are better) or minimize (lower values are better) objective. You may also define a threshold value."

## Contribution

- [x] I'm willing to help implement this feature
- [x] I can help with documentation
- [x] I can help with testing

### PR DESCRIPTION
**Pull request by pkupt**
Implements per-metric **objective** support for the `semantica.evals` runner (issue #1091), aligned with Palantir AIP Evals: each metric can carry a `maximize`/`minimize` direction + threshold, or a Boolean expectation, and a case passes when all its metrics meet their objectives.

Depends on #1090 (the `semantica.evals` module); this branch builds on the runner introduced there.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Related Issues

Closes #1091

## Changes Made

- `evaluate()` now accepts `config={"<evaluator>": {"objective": {"direction": "maximize"|"minimize", "threshold": X}}}` to override the evaluator's default pass verdict with a threshold; `{"objective": {"expect": bool}}` expresses a Boolean expectation
- Runner re-decides each metric's `passed` when an objective is present; no objective → existing behavior unchanged (byte-for-byte)
- Validation (fail-fast, `ValueError` before any evaluator runs): `minimize` requires a `threshold`; `maximize` without one is a no-op; `expect` must be a real bool and cannot combine with `direction`/`threshold`; non-dict objective rejected
- Error metrics are never affected by objectives (error wins over fail)
- `semantica/evals/usage.md` documents the objective config; CHANGELOG entry added
- 14 new tests in `tests/evals/test_runner.py::TestObjective` (76 total in `tests/evals`)

## Testing

- [x] Tested locally
- [x] Added tests for new functionality
- [ ] Package builds successfully (`python -m build`)

### Test Commands

```bash
pytest tests/evals/        # 76 passed
```

## Documentation

- [x] Updated relevant documentation (`semantica/evals/usage.md`)
- [x] Added code examples if applicable (objective examples in `usage.md`)
- [ ] Updated API reference if adding new APIs
- [ ] Updated cookbook if adding new examples

## Breaking Changes

**Breaking Changes**: No — additive, backward-compatible config; `evaluate()` signature and `EvalMetric` shape unchanged.

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Package builds successfully

## Additional Notes

- Design and scope were confirmed by maintainer (@KaifAhmad1 assigned #1091).
- Reference: Palantir AIP Evals — "When configuring an evaluator, you can define an objective for each metric... specify either a maximize (higher values are better) or minimize (lower values are better) objective. You may also define a threshold value."

### CODE REVIEW

**Review by Sameer6305 — APPROVED** [review]
## Final Review — PR #1092

This PR started as the implementation of the **per-metric objective layer for `evaluate()`** from issue #1091. During the review, we went through the implementation, API contract, security implications, evaluator integration, edge cases, tests, documentation, and repository hygiene.

### What we reviewed and found

**1. Objective implementation**

The runner-level objective behavior was verified end-to-end:

* `maximize` + threshold → passes when `score >= threshold`
* `minimize` + threshold → passes when `score <= threshold`
* `maximize` without threshold → preserves evaluator verdict as a no-op
* `minimize` without threshold → must raise `ValueError`
* explicit `threshold=None` follows the same rules
* `expect=True/False` works as an explicit boolean expectation
* non-boolean `expect` values are rejected
* `expect` cannot be combined with `direction`/`threshold`
* invalid objective configuration fails fast before evaluator/target execution
* evaluator errors remain errors and cannot be converted into objective passes
* multiple metrics and cases receive independent objective handling

### Issues found and fixed

**Minimize-without-threshold behavior**

The initial implementation treated `minimize` without a threshold as a no-op. The design spec requires a threshold for `minimize`.

Fixed in **`db93af8`**:

* `minimize` without threshold now raises `ValueError`
* explicit `threshold=None` also raises `ValueError`
* corresponding regression tests were added

**Nested global/per-case objective merging**

We found that a partial per-case objective could replace the entire global objective and accidentally lose fields such as `direction`.

Fixed in **`db93af8`**:

* objective configuration now merges at the nested level
* missing per-case objective fields inherit from global configuration
* configs are not mutated
* conflicting objective forms still fail validation

**Duplicate result models**

`CaseResult` and `EvalSummary` were accidentally defined twice in `types.py`.

Fixed in **`db93af8`** by removing the duplicate definitions while preserving the canonical public models.

**Threshold validation**

We found that some invalid threshold types produced `TypeError` instead of the documented `ValueError`, and non-finite values were not rejected.

Fixed in **`db93af8`**:

* thresholds are coerced consistently
* invalid values raise `ValueError`
* `NaN`, `+inf`, and `-inf` are rejected
* valid numeric values retain existing behavior

We also explicitly reviewed boolean thresholds. The current permissive behavior (`True → 1.0`, `False → 0.0`) is consistent with the project's existing numeric configuration conventions, so no change was necessary.

**Evaluator config / objective passthrough**

We investigated whether the runner should strip the reserved `objective` key before passing evaluator config.

Conclusion: **no change required**. The design spec explicitly documents this passthrough, current evaluators ignore the nested `objective` key, and we found no security or correctness issue.

**Registry test pollution**

We found a test-order dependency where `test_register_and_get` left a dynamically registered evaluator in the global registry, while another test depended on it.

Fixed in **`db93af8`**:

* test registration now uses a dedicated test name
* registry state is cleaned up after each test
* the error-message test no longer depends on test execution order

This was verified with normal, isolated, reverse-order, and duplicate collection.

### Documentation

We also found stale documentation after correcting the `minimize` behavior.

Fixed in **`db93af8`**:

* `usage.md` now states that `minimize` requires a threshold
* `CHANGELOG.md` reflects the same behavior
* stale test count was updated from **76 → 99**

### Validation

The complete eval suite is currently:

**99 passed, 0 failed**

We also verified:

* `git diff --check`
* registry tests independently and in different orders
* no unintended/generated/debug files
* all review-session changes are contained in the intended six files

The full repository suite was attempted but exceeded the 300-second timeout because of unrelated heavy/integration tests; the eval suite itself is completely green.

### Commit history relevant to the review

* **`14d6b1e`** — merge of `main` into the PR branch
* **`db93af8`** — `fix: finalize eval objectives review`
  Contains the review-driven correctness, validation, test-hygiene, and documentation fixes described above.
* **`2c4ba6e`** — latest merge of `main` into the PR branch

### Final assessment

After the review fixes, I found **no remaining correctness, security, API-contract, or objective-behavior blocker**.

The implementation is appropriately isolated in the eval runner, the evaluator contract remains intact, edge cases are covered, and the eval test suite is green.

**Approved from my side.**

File: semantica/evals/runner.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        if direction is not None or threshold is not None:
+            raise ValueError(
+                f"objective for '{name}': 'expect' cannot be combined with "
+                "'direction' or 'threshold'"
+            )
+        return {"expect": expect}
+    if direction == "minimize":
+        if threshold is None:
+            raise ValueError(
+                f"objective for '{name}': 'minimize' requires a 'threshold'"
```
[the remark this answers is not in the document]
Comment (Sameer6305) [reply semantica/evals/runner.py:65]: Resolved in db93af8.

The implementation was corrected to follow the design spec: minimize requires a threshold and raises ValueError when the threshold is omitted or explicitly set to None.

maximize without a threshold remains a no-op, preserving the evaluator's built-in verdict.

Regression tests were added for both omitted and explicit None thresholds.

Verified as part of the 99/99 passing eval test suite.

File: semantica/evals/runner.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                )
+                continue
+        merged = dict(default_config)
+        merged.update(case_config)
+        if expected is None:
+            expected = merged.get("expected")
+        objective_by_name = {
+            name: _parse_objective(name, merged.get(name) or {})
+            for name in evaluators
+        }
```
[the remark this answers is not in the document]
Comment (Sameer6305) [reply semantica/evals/runner.py]: Resolved in db93af8.

Objective configuration is now validated in a dedicated pre-validation pass before any case/target/evaluator execution begins.

This preserves the fail-fast contract: an invalid objective anywhere in the evaluation configuration is rejected before execution starts.

The behavior is covered by regression tests verifying that invalid configuration results in zero target execution.

Verified with the full eval suite: 99/99 passed.

File: semantica/evals/runner.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        if actual is None and resolver is not None:
+            try:
+                actual = resolver(case)
+            except Exception as exc:  # noqa: BLE001
+                case_results.append(
+                    CaseResult(case_id, "error", {}, {"target_fn": str(exc)})
+                )
+                continue
+        merged = dict(default_config)
+        merged.update(case_config)
```
[the remark this answers is not in the document]
Comment (Sameer6305) [reply semantica/evals/runner.py]: Resolved in db93af8.

_merge_config was updated to merge the nested objective configuration as well as the evaluator's top-level configuration.

Per-case objective fields now partially override global objective fields while inheriting unspecified fields from the global objective. Full per-case objectives can still replace the global objective, and conflicting forms continue to fail validation.

The implementation also preserves the no-mutation guarantee for both global and case configuration.

Dedicated merge tests were added and the eval suite is 99/99 passing.

File: semantica/evals/types.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    failed: int
+    errors: int
+    pass_rate: float
+    cases: List[CaseResult] = field(default_factory=list)
+
+
+class CaseResult(NamedTuple):
+    """Evaluation output for a single case."""
+
+    case_id: str
```
[the remark this answers is not in the document]
Comment (Sameer6305) [reply semantica/evals/types.py]: Resolved in db93af8.

The duplicate CaseResult and EvalSummary definitions were removed from semantica/evals/types.py, leaving a single canonical definition of each model.

No public API or runtime behavior was changed by the cleanup.

The complete eval test suite remains 99/99 passing.
